File: cargame/submissions/accepted/cargame_lukas.cpp
/* This is an example solution to the "Car Game" problem from KTH Challenge 2013, held on April 21,
 * 2013.
 *
 * The solution is provided as-is, is not documented, and may be generally difficult to read.
 *
 * This work by KTH CSC (http://www.csc.kth.se/) is licensed under a Creative Commons
 * Attribution-Share Alike 2.5 Sweden License
 * (http://creativecommons.org/licenses/by-sa/2.5/se/deed.en).  You are free to redistribute it as
 * you like as long as this license information is not removed.
 */
//Solution by lukasP (Lukáš Poláček)
#include <iostream>
#include <vector>
#include <string>
#include <cstdio>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)
#define trav(it,c) for(__typeof((c).begin()) it=(c).begin(); it!=(c).end(); ++it)

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
int main()
{
    int n, m; scanf("%d %d", &n, &m);

    vector<string> a(n);
    char c[101];
    rep(i,0,n)
    {
        scanf("%s", c);
        a[i] = c;
    }

    vector<string> b(m);
    rep(i,0,m)
    {
        scanf("%s", c);
        b[i] = c;
    }

    vi ans(m, -1);
    rep(i,0,n)
    {
        vi where(m, 0);
        vector<vi> who(26);
        rep(j,0,m) if (ans[j] == -1)
            who[b[j][0] - 'A'].push_back(j);

        rep(j,0,a[i].size())
        {
            vi oldwho = who[a[i][j] - 'a'];
            who[a[i][j] - 'a'].clear();
            rep(l,0,oldwho.size())
            {
                int w = oldwho[l];
                where[w]++;
                if (where[w] == 3)
                    ans[w] = i;
                else
                    who[b[w][where[w]] - 'A'].push_back(w);
            }
        }
    }

    rep(i,0,m)
    {
        if (ans[i] != -1)
            printf("%s\n", a[ans[i]].c_str());
        else
            printf("No valid word\n");
    }
}

File: cargame/submissions/accepted/cargame_lukas.cs
/* This is an example solution to the "Car Game" problem from KTH Challenge 2013, held on April
 * 21, 2013.
 *
 * The solution is provided as-is, is not documented, and may be generally difficult to read.
 *
 * This work by KTH CSC (http://www.csc.kth.se/) is licensed under a Creative Commons
 * Attribution-Share Alike 2.5 Sweden License
 * (http://creativecommons.org/licenses/by-sa/2.5/se/deed.en).  You are free to redistribute it as
 * you like as long as this license information is not removed.
 */
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

class CarGame
{
    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);

        string[] ss = reader.ReadLine().Split();
        int n = int.Parse(ss[0]), m = int.Parse(ss[1]);

        var a = Enumerable.Range(0, n).Select(x => reader.ReadLine()).ToList();
        var b = Enumerable.Range(0, m).Select(x => reader.ReadLine()).ToList();

        List<int> ans = Enumerable.Repeat(-1, m).ToList();
        List<int> cand = Enumerable.Range(0, m).ToList();
        List<int> invCand = Enumerable.Range(0, m).ToList();
        List<int> where = Enumerable.Repeat(0, m).ToList();

        List<List<int>> who = Enumerable.Range(0, 26).Select(x => new List<int>()).ToList();
        for (int i = 0; i < n; i++)
        {
            foreach (var x in who) x.Clear();
            for (int j = 0; j < cand.Count; j++)
            {
                who[b[cand[j]][0] - 'A'].Add(cand[j]);
                where[j] = 0;
            }

            foreach (char x in a[i])
            {
                List<int> oldwho = new List<int> (who[x - 'a']);
                who[x - 'a'].Clear();
                foreach (int w in oldwho)
                {
                    int pos = invCand[w];
                    if (++where[pos] == 3)
                    {
                        ans[w] = i;
                        // remove w from cand
                        int last = cand.Last();
                        invCand[last] = pos;
                        cand[pos] = last;
                        where[pos] = where.Last();
                        cand.RemoveAt(cand.Count - 1);
                        where.RemoveAt(where.Count - 1);
                    }
                    else
                        who[b[w][where[pos]] - 'A'].Add(w);
                }
            }
        }

        foreach (int x in ans)
            Console.WriteLine(x != -1 ? a[x] : "No valid word");
    }
}
